src/vos/evtree.c

```diff
@@ -2293,61 +2317,41 @@ evt_insert(daos_handle_t toh, const struct evt_entry_in *entry,
 		return rc;
 
 	if (ent_array->ea_ent_nr == 1) {
-		if (entry->ei_rect.rc_minor_epc == EVT_MINOR_EPC_MAX) {
-			/** Special case.   This is an overlapping delete record
-			 *  which can happen when there are minor epochs
-			 *  involved.   Rather than rejecting, insert prefix
-			 *  and/or suffix extents.
+		ent = evt_ent_array_get(ent_array, 0);
+		if (entry->ei_rect.rc_minor_epc == EVT_REBUILD_MINOR_MIN) {
+			ent_cpy                      = *entry;
+			ent_cpy.ei_rect.rc_minor_epc = ent->en_minor_epc + 1;
+			D_ASSERTF(ent_cpy.ei_rect.rc_minor_epc <= EVT_REBUILD_MINOR_MAX,
+				  "minor_epc=%d\n", ent_cpy.ei_rect.rc_minor_epc);
+			/* This should never happen because otherwise, we would
+			 * return no overlap.
 			 */
-			ent = evt_ent_array_get(ent_array, 0);
-			if (ent->en_ext.ex_lo <= entry->ei_rect.rc_ex.ex_lo &&
-			    ent->en_ext.ex_hi >= entry->ei_rect.rc_ex.ex_hi) {
-				/** Nothing to do, existing extent contains
-				 *  the new one
-				 */
-				return 0;
-			}
+			D_ASSERTF(ent_cpy.ei_rect.rc_minor_epc > EVT_REBUILD_MINOR_MIN,
+				  "minor_epc=%d\n", ent_cpy.ei_rect.rc_minor_epc);
+			entryp = &ent_cpy;
+			goto run_tx;
 		}
+		overwrite = true;
 	}
 
+run_tx:
 	rc = evt_tx_begin(tcx);
 	if (rc != 0)
 		return rc;
 
 	if (tcx->tc_depth == 0) { /* empty tree */
-		rc = evt_root_activate(tcx, entry);
+		rc = evt_root_activate(tcx, entryp);
 		if (rc != 0)
 			goto out;
-	} else if (tcx->tc_inob == 0 && entry->ei_inob != 0) {
+	} else if (tcx->tc_inob == 0 && entryp->ei_inob != 0) {
 		rc = evt_root_tx_add(tcx);
 		if (rc != 0)
 			goto out;
-		tcx->tc_inob = tcx->tc_root->tr_inob = entry->ei_inob;
+		tcx->tc_inob = tcx->tc_root->tr_inob = entryp->ei_inob;
 	}
 
 	D_ASSERT(ent_array->ea_ent_nr <= 1);
-	if (ent_array->ea_ent_nr == 1) {
-		if (ent != NULL) {
-			memcpy(&ent_cpy, entry, sizeof(*entry));
-			entryp = &ent_cpy;
-			/** We need to edit the existing extent */
-			if (entry->ei_rect.rc_ex.ex_lo < ent->en_ext.ex_lo) {
-				ent_cpy.ei_rect.rc_ex.ex_hi = ent->en_ext.ex_lo - 1;
-				if (entry->ei_rect.rc_ex.ex_hi <= ent->en_ext.ex_hi)
-					goto insert;
-				/* There is also a suffix, so insert the prefix */
-				rc = evt_insert_entry(tcx, entryp, csum_bufp);
-				if (rc != 0)
-					goto out;
-			}
-
-			D_ASSERT(entry->ei_rect.rc_ex.ex_hi > ent->en_ext.ex_hi);
-			ent_cpy.ei_rect.rc_ex.ex_hi = entry->ei_rect.rc_ex.ex_hi;
-			ent_cpy.ei_rect.rc_ex.ex_lo = ent->en_ext.ex_hi + 1;
-
-			/* Now insert the suffix */
-			goto insert;
```

Comment:
Is this for handling the overlapping removal records? Why it's removed?

Reply:
See above comment.